protocol/l1-upgrades.md

```diff
@@ -0,0 +1,338 @@
+# Purpose
+
+Our current upgrade process is onerous and error prone, requiring a significant amount of effort on
+the part of both the developer and signers. This process is not scalable enough for the needs of
+regular superchain wide upgrades.
+
+This document outlines an improved and standardized approach to upgrading L1 contracts.
+
+## Goals
+
+We take the following as our design goals:
+
+- **Atomicity:** It should be possible to complete an upgrade of the entire superchain's L1
+  contracts within one block.
+- **Declarative design:** As much as possible, a developer should be able to simply indicate what
+  bytecode should be deployed, and what storage values (if any) should be initialized or modified.
+- **Deterministic upgrade calldata:** The exact calldata to be submitted during the upgrade should
+  predictable at the time of the governance proposal.
+- **Safety:** Above all the upgrade must be safe, particularly with respect to the things which can
+  typically go wrong with an upgrade, ie.
+  - All contracts should be upgraded to the correct implementation.
+  - Storage layouts should not be incorrectly modified.
+  - Contracts should not be 're-initializable'
+- **Simplicity:** We avoid trying to account for all special cases, and instead require contracts to
+  be implemented in a manner compatible with the solution outlined here.
+- **Auditability:** An upgrade path should be easy to understand, either by reviewing the source
+  code as in the case of a forthcoming upgrade, or by reviewing the on-chain history along with
+  verified source code.
+
+## Non-goals
+
+This document is not concerned with:
+
+1. upgrading predeploys or other L2 contracts.
+1. modifications to the ownership system, ie, it does not attempt to modify Safes, modules or
+   guards.
+
+More over, it does not present a plan for managing contracts which are not proxied, and not
+Multichain Prepared (MCP-L1) contracts (ie. FaultDisputeGame, AnchorStateRegistry). This
+work will either include, or else depend on, all L1 contracts becoming either:
+
+1. Proxied and MCP ready
+2. Superchain-shared contracts (proxied or not)
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+The proposed solution will perform upgrades by transferring upgrade authorization of all L1
+contracts to the release specific OPCM. Upon completion of the upgrade, this authorization will
+be returned to the Security Council.
+
+Rather than the curent `StorageSetter` based approach which simply deletes the `_initialized` slot,
+allowing for `initialize()` to be called, we take this two-step upgrade further, with a short-lived
+`Populator` contract being temporarily set as the implementation in order to set or modify any
```

Comment:
I do not buy the benefits listed here as the benefits of the `Populator` contract. The `Populator` contract is basically `StorageSetter` but with awareness of the storage layout of the particular contract. You can do everything you would with `Populator` with `StorageSetter`. It could be the case that its just more simple to continue to use `StorageSetter` (and improve its API if necessary with things like `setBool(bool _value, uint256 _offset)`)

The benefits listed are benefits, but have nothing to do with `Populator` specifically

Reply:
I listed the benefits [here](https://github.com/ethereum-optimism/design-docs/pull/159/files/34679b374bbf155a64182e5527b58f56a6c07faf#r1850718398). 

Question for you: 

IMO this pattern is similar to the `OptimismPortalMigration` you describe [here](https://github.com/ethereum-optimism/design-docs/issues/154), and I'd have used the `Migration` rather than `Populator` name had I seen that issue before writing this doc. 

Could you please enumerate the benefits you think my proposal is missing from yours, so that I can try to address them in identifying the final design?

Reply:
Two things:
- Base contract with storage layout inherited by the impl and the migrator rather than extending the impl with `initialize()`
- I now believe its less work to just use `StorageSetter` and hand write setting the storage slots in the impl of OPCM. This is very easy to cover with testing and reduces a ton of boilerplate that needs to be tested (every populator contract having a test) as well as boilerplate that needs to go on chain (why does populator have the impl details of the contract?)

Reply:
Resolving. Please leave feedback on that plan on the "Modfications to L1 contracts" section.